Update Working with Dependencies -> Declaring Repositories Section (#30525)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/dep-man/declaring-repositories/supported_repository_types.adoc

// Copyright (C) 2023 Gradle, Inc.
//
// Licensed under the Creative Commons Attribution-Noncommercial-ShareAlike 4.0 International License.;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://creativecommons.org/licenses/by-nc-sa/4.0/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[repository-types]]
= Repository Types

> Gradle supports various sources for resolving dependencies, accommodating different formats and connectivity methods.

Author: Done.

> You can resolve dependencies from:

Author: Done.

- <<#sub:flat_dir_resolver,Local (flat) directories>>
- <<#sec:maven_repo,Maven-compatible artifact repositories>> (e.g., Maven Central)
- <<#sec:ivy_repositories,Ivy-compatible artifact repositories>> (including custom layouts)

[[sub:flat_dir_resolver]]
== Flat directory repository

Some projects store dependencies on a shared drive or within the project’s source code rather than using a binary repository.
To use a flat filesystem directory as a repository, you can configure it like this:

====
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=flat-dir-multi]"]
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=flat-dir-multi]"]
====

This configuration adds repositories that search specified directories for dependencies.

> Flat directory repositories do not support metadata formats like Ivy XML or Maven POM files.

Reviewer: ```suggestion
NOTE: Flat directory repositories are generally discouraged, as they do not support metadata formats like Ivy XML or Maven POM files.
In general, binary dependencies should be sourced from an external repository, but if storing dependencies externally is not an option, prefer declaring a Maven or Ivy repository using a local file URL instead.
```

Author: Done.

> Instead, Gradle dynamically generates a module descriptor based on the presence of artifacts, without any dependency information.
> 
> Gradle prefers modules with real metadata over those generated by flat directory repositories, so flat directories cannot override artifacts with real metadata from other declared repositories.
> For instance, if Gradle finds `jmxri-1.2.1.jar` in a flat directory and `jmxri-1.2.1.pom` in another repository, it will use the metadata from the latter.

Reviewer: ```suggestion
When resolving dependencies from a flat dir repo, Gradle dynamically generates adhoc dependency metadata based on the presence of artifacts.
Gradle prefers modules with real metadata over those generated by flat directory repositories. For this reason, flat directories cannot override artifacts with real metadata from other declared repositories.

For instance, if Gradle finds `jmxri-1.2.1.jar` in a flat directory and `jmxri-1.2.1.pom` in another repository, it will use the metadata from the latter.
```

Author: Done.

> If you need to override remote artifacts with local ones, consider using an Ivy or Maven repository pointing to a local directory instead.

Reviewer: ```suggestion
```

Deleted and moved above

Author: Removed.

[[sec:maven_repo]]
== Maven repositories

> Many organizations host dependencies in private Maven repositories accessible within their network.

Reviewer: These APIs can also be used for public Maven repos, not only private repos

Author: Fixed.

Gradle can declare Maven repositories by specifying their URL:

====
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=maven-like-repo]"]
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=maven-like-repo]"]
====

[[sub:custom-maven-repo]]

> === Composite Maven repository

Reviewer: TIL This exists.

Why.... Who does this?

Author: I didn't want to eliminate it since someone clearly added this for a reason. It is now removed.

Reviewer: I was more being rhetorical here. I am just surprised this feature exists. 

As long as it does exist I think it makes sense to document. This may just by on my chopping block a little sooner now that I know about it

Sometimes, POMs are published in one location, and JARs in another.
You can define such a repository as follows:

====
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=maven-like-repo-with-jar-repo]"]
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=maven-like-repo-with-jar-repo]"]
====

Gradle will first look for POMs and artifacts at the base URL, and if the artifact is not found, it will check the additional `artifactUrls`.

[[sub:accessing_secured_maven_repositories]]
=== Authenticated Maven repository

You can specify credentials for Maven repositories that require authentication.
See <<supported_repository_protocols.adoc#supported_transport_protocols>> for authentication options.

[[sub:maven_local]]
=== Local Maven repository

> Gradle can also consume dependencies from the link:https://maven.apache.org/guides/introduction/introduction-to-repositories.html[local Maven repository].
> This is useful for teams that publish locally with one project and consume the artifacts with another.
> To declare the local Maven cache as a repository:

Reviewer: ```suggestion
Gradle can also consume dependencies from the link:https://maven.apache.org/guides/introduction/introduction-to-repositories.html[local Maven repository]:
```

This pattern should really be discouraged and is only really intended for integration with Maven.

Author: Simplified and discouraged.

====
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=maven-local]"]
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=maven-local]"]
====

NOTE: Gradle stores resolved dependencies in <<dependency_resolution.adoc#sec:dependency_cache,its own cache>>.
A build does not need to declare the local Maven repository even if you resolve dependencies from a Maven-based, remote repository.

WARNING: Before adding Maven local as a repository, you should <<declaring_repositories_basics.adoc#sec:case-for-maven-local, make sure this is really required>>.

Gradle uses the same logic as Maven to determine the location of your local Maven cache.
It prioritizes the `settings.xml` in the user’s home directory,`<user_home>/.m2`, over the one in `__M2_HOME__/conf`.

> If no `settings.xml` is found, Gradle uses the default location at `<user_home>/.m2/repository`.

Reviewer: We've already described a lot of this info on the other page. I'd recommend we avoid duplicating it and either link from that page to this or the other way around

Author: Removed.

[[sec:ivy_repositories]]
== Ivy repositories

> Organizations may choose to host dependencies in an in-house Ivy repository.

Reviewer: These APIs can be used for public Ivy repos, not just in-house repos (though I'm not sure in practice there are many public ivy repos out there)

Author: Fixed.

=== Custom pattern layout Ivy repository

To define an Ivy repository with a non-standard layout, you can set up a pattern layout:

====
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo-with-pattern-layout]"]
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=ivy-repo-with-pattern-layout]"]
====

For an Ivy repository that fetches Ivy files and artifacts from different locations, define separate patterns:

====
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo-with-custom-pattern]"]
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=ivy-repo-with-custom-pattern]"]
====

Optionally, you can enable Maven-style layout for the 'organisation' part, with forward slashes replacing dots:

====
include::sample[dir="snippets/artifacts/defineRepository/kotlin",files="build.gradle.kts[tags=ivy-repo-with-m2compatible-layout]"]
include::sample[dir="snippets/artifacts/defineRepository/groovy",files="build.gradle[tags=ivy-repo-with-m2compatible-layout]"]
====

[[sub:accessing_secured_ivy_repositories]]

> === Authenticated Ivy repository

Reviewer: I find it odd that we link to another page for maven authentication but include ivy authentication on this page

Author: Everything has been moved to the supported_repo_types page.
